### ISSUE TITLE
Make longFormat a generic

### ISSUE THREAD

**Issue by lgatto** [waldronlab/MultiAssayExperiment#333]
Would you consider making `longFormat()` a generic function (ideally in `BiocGenerics`) to allow implementation if different methods?

Currently, you have 

```r
MultiAssayExperiment::longFormat
function (object, colDataCols = NULL, i = 1L) 
{
    if (is(object, "ExperimentList")) 
        return(do.call(rbind, .longFormatElist(object, i = i)))
    else if (!is(object, "MultiAssayExperiment")) 
        return(.longFormatANY(object, i = i))
    if (any(.emptyAssays(experiments(object)))) 
        object <- .dropEmpty(object, warn = FALSE)
    longDataFrame <- do.call(function(...) rbind(..., make.row.names = FALSE), 
        .longFormatElist(experiments(object), i = i))
    longDataFrame <- .mapOrderPrimary(longDataFrame, sampleMap(object))
    if (!is.null(colDataCols)) 
        longDataFrame <- .matchAddColData(longDataFrame, colData(object), 
            colDataCols)
    as(longDataFrame, "DataFrame")
}
<bytecode: 0x5c13c5da9e48>
<environment: namespace:MultiAssayExperiment>
```

The `QFeatures` class extends `MultiAssayExperiment`, and has 

```r
QFeatures::longFormat
function (object, colvars = NULL, rowvars = NULL, index = 1L) 
{
    if (!is.null(rowvars)) {
        rdNames <- rowDataNames(object)
        misNames <- sapply(rdNames, function(x) any(!rowvars %in% 
            x))
        if (any(misNames)) 
            stop("Some 'rowvars' not found in assay(s): ", paste0(names(misNames)[misNames], 
                collapse = ", "))
        longDataFrame <- MultiAssayExperiment::longFormat(object, 
            colvars, index)
        rds <- lapply(rowData(object), function(rd) rd[, rowvars, 
            drop = FALSE])
        rds <- do.call(rbind, rds)
        cbind(longDataFrame, rds[as.character(longDataFrame$rowname), 
            , drop = FALSE])
    }
    else {
        MultiAssayExperiment::longFormat(object, colvars, index)
    }
}
<bytecode: 0x5c13df3029a0>
<environment: namespace:QFeatures>
```

The main advantage is support for variables from `colData` and `rowData` to the long format output (which is very handy). 

The main motivation for this request is that I would also like to be able to convert `SummarizedExperiment` objects to long format.

**Comment by lgatto** [comment]
- Now that I look into it more carefully, `MultiAssayExperiment::longFormat()` isn't a method, but works on SEs calling `MultiAssayExperiment:::.longFormatANY()`. What is the reason for not using methods to handle this?
- And my second/other request would be to support `rowData` variables.

**Comment by LiNk-NY** [comment]
Hi Laurent, @lgatto 
Sorry, I just saw your issue. 
When I wrote `longFormat`, I didn't think it would be useful for other classes so I didn't create a generic / method. 
I could create the generic and methods and keep the `ANY` method to work with other classes.

**Comment by lgatto** [comment]
I find it very useful, as it is a seamless transition to ggplot visualisation.

>  I could create the generic and methods and keep the ANY method to work with other classes.

I think a generic in `BiocGenerics` would be great, dispatching on `object` only, with `...` to provide some flexibility to add other arguments. 

Once a generic becomes available, I would implement methods for `QFeatures` and `SummarizedExperiment` (I could also contribute the latter to the SE package, if the maintainer find it useful).

Thanks!

**Comment by LiNk-NY** [comment]
Hi Laurent, @lgatto 

I added this to the devel branch. I don't know if the `longFormat` generic would be good in `BiocGenerics` for now it is in `MultiAssayExperiment`. 

Thanks!

**Comment by lgatto** [comment]
OK, thank you. 

At this point, having the generic in MAE isn't a problem for me as my package/class extends MAE, and hence I depend on it anyway. But I think it is a limitation for other cases.

**Comment by lgatto** [comment]
Looking at the generic [signature](https://github.com/waldronlab/MultiAssayExperiment/blob/devel/R/MultiAssayExperiment-helpers.R#L350), do you really need `colDataCols` and `i` in there? Isn't this restrictive for a broader usage? Do you see a limitation in defining the generic as

```r
setGeneric(
    "longFormat",
    function(object, ...)
    standardGeneric("longFormat")
)

```

**Comment by lgatto** [comment]
Yes, all arguments don't have to be used in a method, and new arguments can also be added. An issue with arguments that are not used in methods is that a user could still write them, and they would simply get ignored in a method that doesn't use them. They can also be in the way due to partial matching. This is, as far as I understand, (some of) the reasons to keep generics with minimal signatures, and define arguments that would be needed for *all anticipated* methods. 

I understand that you would prefer to keep *i*. It indeed makes sense for MAE and SE objects, and the default value of 1 would also make sense when it isn't relevant at all (not *i* or the first *i* is conceptually similar enough). For `colDataCols`, I think it is already quite specialised (i.e. objects that have a `colData`) and could be left for the methods to implement/use.

**Comment by LiNk-NY** [comment]
IMO partial matching should be a bug not a feature. I recommend these partial match `options` listed by Kevin Ushey here: https://kevinushey.github.io/blog/2015/02/02/rprofile-essentials/ 
It is hard to anticipate all methods, e.g., from methods in `SummarizedExperiment` to `SpatialExperiment`. 

I was concerned about the function signature changing (when converted to a generic) and having to deprecate the `colDataCols` argument but it seems that that won't be necessary if methods and generics are documented in the same man page.

**Comment by LiNk-NY** [comment]
It seems that moving the argument would present a problem for existing code: 

https://github.com/waldronlab/MultiAssayExperiment/pull/335/files

The signature `function(object, colDataCols = NULL, i = 1L)` would change to `function(object, i = 1L)` putting `i` in the second position and breaking any existing code that uses `colDataCols`. There removal of the argument would have to be done more systematically.

**Comment by lgatto** [comment]
Thank you Marcel. 

> IMO partial matching should be a bug not a feature.

Yes, I agree. I do warn on partial matching, but we can't count on all users doing so. It is a constant risk indeed.

**Comment by lgatto** [comment]
> Is this close to what you have in mind for the longFormat() method for SummarizedExperiment objects?

Yes, it looks like very similar to what I already do (for multiple SEs in a MAE-like object), except that a `DataFrame` is returned, not printed. It would indeed be useful to harmonise these operations across packages.

**Comment by LiNk-NY** [comment]
Hi Hervé, @hpages 
Thanks for the suggestion!
I am in favor of renaming it to `longForm` and creating a generic in `BiocGenerics`.

**Comment by hpages** [comment]
@LiNk-NY @lgatto I just added `longForm()` to **BiocGenerics** 0.53.4. See https://github.com/Bioconductor/BiocGenerics/commit/ad956f215ca7441152a962ba56b664975967a318

**Comment by LiNk-NY** [comment]
Thanks Hervé!
@lgatto 
I added this to **MultiAssayExperiment** version `1.33.6` https://github.com/waldronlab/MultiAssayExperiment/commit/a42ee78ee89f7572e6a4ce6da6ea6e9b50c2026c.

**Comment by lgatto** [comment]
Thank you both. I'll implement the methods downstream from MAE
.
